### PR TITLE
Add TOML cheatcode support

### LINKED ISSUE #4808: feat (cheatcode) TOML support
### Component

Forge

### Describe the feature you would like

The TOML format is used by foundry itself and is quite an elegant configuration format.

it would be great if we could have TOML parsing facility like we do for json via vm cheatcode

This would allow us to also read the foundry toml file and this way we could extend it with new parameters 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by zerosnacks**
## Motivation

Closes #4808 

Implements TOML cheatcode support equivalent to JSON cheatcode support

## Solution

The implementation is a thin serialisation / deserialisation layer to convert between TOML to JSON. Internally it re-uses the existing `serialized_jsons` state. Has feature parity with JSON cheatcodes people are already familiar with.

Implements the following new cheatcodes:

- `parseToml(string)`
- `parseToml(string,key)` (`key` support!)
- All coercion cheatcodes (e.g. `parseTomlUint(string,key)`).
- `writeToml(string,path)`
- `writeToml(string,path,key)`

Re-uses JSON's non-prefixed `serialize*` methods as they operate on the `serialized_jsons` state.

Special case is: `keyExistsToml(string,key)` of which the equivalent is `keyExists(string,key)` in JSON (introduced in https://github.com/foundry-rs/foundry/pull/5431). This was chosen to introduce no breaking changes.

Whilst this approach has some (minor) overhead it was chosen to limit complexity, increase maintainability, avoid breaking changes and achieve feature parity (including paths through `jsonpath_lib` of which no equivalent TOML library exists).

The test suite is a modified version of `Json.t.sol`. A possible improvement could be a shared abstraction of the two tests but I consider that to be out of scope for this PR.

At-merge PR:
- [ ] Documentation in `foundry-book`: https://github.com/foundry-rs/book/pull/1133

Post-merge PR:
- [ ] Add upstream support to `forge-std` for the proposed cheatcodes: https://github.com/foundry-rs/forge-std/pull/518

Known limitations:
- Currently the conversion from JSON `string` to TOML `datetime` is not handled as it would require a Regex match in a likely hot path for something that is highly uncommon.

### CODE REVIEW

File: crates/cheatcodes/spec/src/cheatcode.rs [hunk 1]
```diff
@@ -108,6 +108,12 @@ pub enum Group {
     ///
     /// Safety: safe.
     Json,
+    /// Utility cheatcodes that deal with parsing values from and converting values to TOML.
+    ///
+    /// Examples: `serializeToml`, `parseToml`, `writeToml`.
```
Comment (zerosnacks) [inline crates/cheatcodes/spec/src/cheatcode.rs:113]: To do: remove `serializeToml`, not implemented

File: crates/cheatcodes/src/toml.rs [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+fn convert(toml: &str) -> Result<String> {
+    let toml = parse_toml_str(toml)?;
+    let json = toml_to_json(toml);
+    serde_json::to_string(&json).map_err(|e| fmt_err!("failed to convert to JSON: {e}"))
+}
+
+/// Convert a TOML value to a JSON value.
+fn toml_to_json(value: TomlValue) -> JsonValue {
+    match value {
+        TomlValue::String(s) => JsonValue::String(s),
```
Comment (DaniPopes) [inline crates/cheatcodes/src/toml.rs:188]: do we want to convert `"null"` to `JsonValue::Null` here?

File: crates/cheatcodes/src/toml.rs [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    fn apply(&self, _state: &mut Cheatcodes) -> Result {
+        let Self { toml, key } = self;
+        parse_json_coerce(&convert(toml)?, key, &DynSolType::FixedBytes(32))
+    }
+}
+
+impl Cheatcode for parseTomlKeysCall {
+    fn apply(&self, _state: &mut Cheatcodes) -> Result {
+        let Self { toml, key } = self;
+        parse_json_keys(&convert(toml)?, key)
```
Comment (DaniPopes) [inline crates/cheatcodes/src/toml.rs:139]: We can dedup the `parse_json(convert()?)` to their own parse_toml functions, wdyt?

File: crates/cheatcodes/src/toml.rs [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+}
+
+fn parse_toml_str(toml: &str) -> Result<TomlValue> {
+    toml::from_str(toml).map_err(|e| fmt_err!("failed parsing TOML: {e}"))
+}
+
+/// Convert a TOML string to a JSON string.
+fn convert(toml: &str) -> Result<String> {
+    let toml = parse_toml_str(toml)?;
+    let json = toml_to_json(toml);
```
Comment (DaniPopes) [inline crates/cheatcodes/src/toml.rs:181]: Would it be possible to avoid this conversion by just using the `Serialize` implementation on `TomlValue`? https://docs.rs/toml/latest/src/toml/value.rs.html#398-443

**Review by DaniPopes — COMMENTED** [review]
Approach LGTM, just a few questions.

File: crates/cheatcodes/spec/src/vm.rs [hunk 5]
```diff
@@ -1816,6 +1816,98 @@ interface Vm {
...
+    // ======== TOML Parsing and Manipulation ========
+
+    // -------- Reading --------
+
+    // NOTE: Please read https://book.getfoundry.sh/cheatcodes/parse-toml to understand the
+    // limitations and caveats of the TOML parsing cheat.
+
+    /// Checks if `key` exists in a TOML table.
+    #[cheatcode(group = Toml)]
+    function keyExistsToml(string calldata toml, string calldata key) external view returns (bool);
```
[the remark this answers is not in the document]
Comment (zerosnacks) [reply crates/cheatcodes/spec/src/vm.rs:1828]: Good point, I'm in favor.

I don't see an elegant way of reusing the `keyExists` method without introducing brittle complexity (like a Regex). Would rather opt for explicit naming.

I did do a brief search on Github to see the impact and it should be relatively limited: https://github.com/search?q=vm.keyExists&type=code but we would still need an elegant way of issuing a deprecation message and schedule a removal. I'm unfamiliar with that process.

Happy to add it to this PR or open a separate one.

[the remark this answers is not in the document]
Comment (mds1) [reply crates/cheatcodes/spec/src/vm.rs:1828]: I'd say we make the change in this PR. In the past we've used console.logs to surface warnings about cheats being deprecated. That was in forge-std though, presumably we can do the same thing here by inserting the required call to the console address in the cheat implementation. But will defer to @DaniPopes @klkvr and co. on how they prefer to do this

File: crates/cheatcodes/src/toml.rs (hunk 4 above)
Comment (zerosnacks) [reply crates/cheatcodes/src/toml.rs]: Good point! I initially ran into issues with the direct conversion, specifically around arrays and objects, but those turned out to be unrelated. I've updated the implementation to use direct conversions, much nicer.

File: crates/cheatcodes/spec/src/vm.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (zerosnacks) [reply crates/cheatcodes/spec/src/vm.rs:1832]: I've opted for adding the `keyExistsJson` alias and marked `keyExists` with the `Deprecated` status. I've also added a comment in a few places marking the deprecation.

I've added a bug report for the emitting of warnings and errors here: https://github.com/foundry-rs/foundry/issues/7323

File: crates/evm/traces/src/decoder/mod.rs [hunk 6]
```diff
@@ -441,12 +442,46 @@ impl CallTraceDecoder {
...
+                    {
+                        "<JSON file>"
+                    } else {
+                        "<stringified JSON>"
+                    };
+                    decoded[0] = DynSolValue::String(token.to_string());
+                    Some(decoded.iter().map(format_token).collect())
+                }
+            }
+            "parseToml" |
```
[the remark this answers is not in the document]
Comment (zerosnacks) [reply crates/evm/traces/src/decoder/mod.rs]: Done

File: crates/cheatcodes/src/toml.rs (hunk 2 above)
Comment (zerosnacks) [reply crates/cheatcodes/src/toml.rs]: For simplicity I've decided to rely on the default Serde behavior since removing the custom handling of the TOML -> JSON value conversion. The default behavior is to cast it to `None` and ignore the key. Internally the JSON `null` is converted to a `bytes32(0)` (but to my understanding not converted back upon writing).

I've documented the edge case here: https://github.com/foundry-rs/book/pull/1133/files#r1514559024

Comment (zerosnacks) [reply crates/cheatcodes/src/toml.rs]: Done, handles conversion to and from now.

File: crates/cheatcodes/src/toml.rs [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+                Some(f) => TomlValue::Float(f),
+                None => TomlValue::String(n.to_string()),
+            },
+        },
+        JsonValue::Bool(b) => TomlValue::Boolean(b),
+        JsonValue::Array(a) => TomlValue::Array(a.into_iter().map(json_to_toml_value).collect()),
+        JsonValue::Object(o) => {
+            TomlValue::Table(o.into_iter().map(|(k, v)| (k, json_to_toml_value(v))).collect())
+        }
+        JsonValue::Null => TomlValue::String("null".to_owned()),
```
Comment (mattsse) [inline crates/cheatcodes/src/toml.rs:241]:
```suggestion
        JsonValue::Null => TomlValue::String("null".to_string()),
```

File: testdata/fixtures/Toml/write_complex_test.toml [hunk 8]
```diff
@@ -0,0 +1,6 @@
+a = 123
+b = "test"
+
+[c]
+a = 123
+b = "test"
```
Comment (mattsse) [inline testdata/fixtures/Toml/write_complex_test.toml:6]: do we have a test for this table?

**Review by mattsse — COMMENTED** [review]
test look good!

ptal at cheatcode impl @DaniPopes @klkvr

File: crates/cheatcodes/src/json.rs [hunk 9]
```diff
@@ -431,6 +446,10 @@ pub(super) fn value_to_token(value: &Value) -> Result<DynSolValue> {
                     32 => DynSolValue::FixedBytes(B256::from_slice(&bytes), 32),
                     _ => DynSolValue::Bytes(bytes),
                 })
+            } else if string == "true" {
+                Ok(DynSolValue::Bool(true))
+            } else if string == "false" {
+                Ok(DynSolValue::Bool(false))
```
Comment (DaniPopes) [inline crates/cheatcodes/src/json.rs:452]: Why was this added?

**Review by DaniPopes — APPROVED** [review]
LGTM

File: crates/cheatcodes/src/toml.rs (hunk 7 above)
Comment (zerosnacks) [reply crates/cheatcodes/src/toml.rs]: Thanks!

File: testdata/fixtures/Toml/write_complex_test.toml (hunk 8 above)
Comment (zerosnacks) [reply testdata/fixtures/Toml/write_complex_test.toml:6]: Yes we do here, the test specifically covers the writing aspect: https://github.com/foundry-rs/foundry/blob/493c6153d0d55bb747b24f179635e7690585b176/testdata/cheats/Toml.t.sol#L259-L270

File: crates/cheatcodes/src/json.rs (hunk 9 above)
Comment (zerosnacks) [reply crates/cheatcodes/src/json.rs]: Good catch, I added it to implicitly handle bool conversion in the same way we handle the `null` case when I was improving the test suite. Looking at it again it is better to not implement it and instead defer to the original implementation where users can use type coercion to cast this to a bool.

I've made some improvements to the boolean tests of JSON and TOML to make sure the existing implementation is correct.

**Review by mattsse — APPROVED** [review]
lgtm
send it @zerosnacks
